File: python/vierrataleai/providers/openai.py
import json
import urllib.request
import urllib.error
from typing import AsyncGenerator, List, Optional

from .base import BaseProvider
from ..catalog import get_real_model, get_display_name, is_local_model, is_cloud_model
from .. import config


class OpenAIProvider(BaseProvider):
    def __init__(self):
        super().__init__("openai")

    async def is_available(self) -> bool:
        key = config.get("openai_api_key")
        if not key:
            return False
        try:
            req = urllib.request.Request(
                "https://api.openai.com/v1/models",
                headers={"Authorization": f"Bearer {key}"},
            )
            resp = urllib.request.urlopen(req, timeout=5)
            return resp.status == 200
        except Exception:
            return False

    async def list_models(self) -> List[dict]:
        key = config.get("openai_api_key")
        if not key:
            return []
        try:
            req = urllib.request.Request(
                "https://api.openai.com/v1/models",
                headers={"Authorization": f"Bearer {key}"},
            )
            resp = urllib.request.urlopen(req, timeout=10)
            data = json.loads(resp.read())
            return [
                {
                    "real_name": m["id"],
                    "display_name": get_display_name(m["id"]),
                }
                for m in data.get("data", [])
                if m["id"].startswith("gpt-")
            ]
        except Exception:
            return []

    def _resolve_model(self, model_name: str) -> str:
        if is_local_model(model_name) or not is_cloud_model(model_name):
            return "gpt-4o-mini"
        return get_real_model(model_name)

    async def stream(
        self,
        messages: List[dict],
        options: Optional[dict] = None,
    ) -> AsyncGenerator[str, None]:
        options = options or {}
        key = config.get("openai_api_key")
        if not key:
            raise RuntimeError("[ERR-0004] Cloud API key not configured. Add it to the config and try again.")

        model = self._resolve_model(options.get("model", config.get("model")))
        system_prompt = options.get("system_prompt", "")

        api_messages = []
        if system_prompt:
            api_messages.append({"role": "system", "content": system_prompt})
        for msg in messages:
            api_messages.append({"role": msg["role"], "content": msg["content"]})

        body = json.dumps({
            "model": model,
            "messages": api_messages,
            "stream": True,
            "temperature": options.get("temperature", config.get("temperature")),
            "max_tokens": options.get("max_tokens", config.get("max_tokens")),
        }).encode()

        req = urllib.request.Request(
            "https://api.openai.com/v1/chat/completions",
            data=body,
            headers={
                "Content-Type": "application/json",
                "Authorization": f"Bearer {key}",
            },
        )

        try:
            resp = urllib.request.urlopen(req, timeout=240)
        except urllib.error.HTTPError as err:
            raise RuntimeError(
                f"[ERR-0003] The cloud provider returned HTTP {err.code}. Check your API key in config."
            )
        except Exception as err:
            raise RuntimeError("[ERR-0001] Could not reach the API (no response within 120s). Check your network connection.")

        buffer = ""
        for chunk in iter(lambda: resp.read(1024), b""):
            buffer += chunk.decode("utf-8", errors="ignore")
            lines = buffer.split("\n")
            buffer = lines.pop() or ""

            for line in lines:
                if not line.startswith("data: "):
                    continue
                data_str = line[6:]
                if data_str == "[DONE]":
                    return
                try:
                    data = json.loads(data_str)
                    content = data.get("choices", [{}])[0].get("delta", {}).get("content", "")
                    if content:
                        yield content
                except json.JSONDecodeError:
                    continue

File: nodejs/src/providers/openai.js
import { BaseProvider } from './base.js';
import { Catalog } from '../catalog.js';
import { Config } from '../config.js';

export class OpenAIProvider extends BaseProvider {
  constructor() {
    super('openai');
  }

  get displayName() {
    return 'Nebula';
  }

  async isAvailable() {
    const key = Config.get('openaiApiKey');
    if (!key) return false;
    try {
      const resp = await fetch('https://api.openai.com/v1/models', {
        headers: { Authorization: `Bearer ${key}` },
        signal: AbortSignal.timeout(5000),
      });
      return resp.ok;
    } catch {
      return false;
    }
  }

  async listModels() {
    const key = Config.get('openaiApiKey');
    if (!key) return [];
    try {
      const resp = await fetch('https://api.openai.com/v1/models', {
        headers: { Authorization: `Bearer ${key}` },
      });
      if (!resp.ok) return [];
      const data = await resp.json();
      return (data.data || [])
        .filter((m) => m.id.startsWith('gpt-'))
        .map((m) => ({
          realName: m.id,
          displayName: Catalog.getDisplayName(m.id),
        }));
    } catch {
      return [];
    }
  }

  // If the configured model is a local (ollama) model, use a cloud default so
  // we never send a qwen model name to the OpenAI API.
  _resolveModel(options) {
    const requested = options.model || Config.get('model');
    if (Catalog.isLocalModel(requested) || !Catalog.isCloudModel(requested)) {
      return 'gpt-4o-mini';
    }
    return Catalog.getRealModel(requested);
  }

  async *stream(messages, options = {}) {
    const key = Config.get('openaiApiKey');
    if (!key) throw new Error('[ERR-0004] Cloud API key not configured. Add it to the config and try again.');

    const model = this._resolveModel(options);
    const systemPrompt = options.systemPrompt || '';

    const apiMessages = [];
    if (systemPrompt) {
      apiMessages.push({ role: 'system', content: systemPrompt });
    }
    for (const msg of messages) {
      apiMessages.push({ role: msg.role, content: msg.content });
    }

    const controller = new AbortController();
    const connectTimer = setTimeout(() => controller.abort(), 240000);

    let resp;
    try {
      resp = await fetch('https://api.openai.com/v1/chat/completions', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          Authorization: `Bearer ${key}`,
        },
        body: JSON.stringify({
          model,
          messages: apiMessages,
          stream: true,
          temperature: options.temperature || Config.get('temperature'),
          max_tokens: options.maxTokens || Config.get('maxTokens'),
        }),
        signal: controller.signal,
      });
    } catch (err) {
      if (err.name === 'AbortError') {
        throw new Error('[ERR-0001] Could not reach the API (no response within 60s). Check your network connection.');
      }
      throw new Error('[ERR-0001] Could not reach the API (no response). Check your network connection.');
    } finally {
      clearTimeout(connectTimer);
    }

    if (!resp.ok) {
      throw new Error('[ERR-0003] The cloud provider returned HTTP ' + resp.status + '. Check your API key in config.');
    }

    const reader = resp.body.getReader();
    const decoder = new TextDecoder();
    let buffer = '';
    let idle = null;
    const settle = () => {
      if (idle) clearTimeout(idle);
      idle = setTimeout(() => controller.abort(), 240000);
    };
    settle();
    try {
      while (true) {
        let value, done;
        try {
          ({ done, value } = await reader.read());
        } catch (err) {
          if (err.name === 'AbortError') {
            throw new Error('[ERR-0001] Lost the connection to the cloud API (no data for 60s).');
          }
          throw err;
        }
        if (done) break;
        settle();
        buffer += decoder.decode(value, { stream: true });
        const lines = buffer.split('\n');
        buffer = lines.pop() || '';

        for (const line of lines) {
          if (!line.startsWith('data: ')) continue;
          const data = line.slice(6);
          if (data === '[DONE]') return;
          try {
            const json = JSON.parse(data);
            const content = json.choices?.[0]?.delta?.content;
            if (content) yield content;
          } catch {}
        }
      }
    } finally {
      if (idle) clearTimeout(idle);
    }
  }
}
